// react-navigation-material-bottom-tabs/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createMaterialBottomTabNavigator", {
  enumerable: true,
  get: function () {
    return _createMaterialBottomTabNavigator.default;
  }
});
Object.defineProperty(exports, "NavigationTabState", {
  enumerable: true,
  get: function () {
    return _types.NavigationTabState;
  }
});
Object.defineProperty(exports, "NavigationTabProp", {
  enumerable: true,
  get: function () {
    return _types.NavigationTabProp;
  }
});
Object.defineProperty(exports, "NavigationTabScreenProps", {
  enumerable: true,
  get: function () {
    return _types.NavigationTabScreenProps;
  }
});
Object.defineProperty(exports, "NavigationMaterialBottomTabOptions", {
  enumerable: true,
  get: function () {
    return _types.NavigationMaterialBottomTabOptions;
  }
});
Object.defineProperty(exports, "NavigationMaterialBottomTabConfig", {
  enumerable: true,
  get: function () {
    return _types.NavigationMaterialBottomTabConfig;
  }
});
Object.defineProperty(exports, "NavigationMaterialBottomTabScreenComponent", {
  enumerable: true,
  get: function () {
    return _types.NavigationMaterialBottomTabScreenComponent;
  }
});

var _createMaterialBottomTabNavigator = _interopRequireDefault(require("./navigators/createMaterialBottomTabNavigator"));

var _types = require("./types");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
