// react-navigation-material-bottom-tabs/lib/module/views/MaterialBottomTabView.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as React from 'react';
import { ThemeContext } from 'react-navigation';
import { BottomNavigation } from 'react-native-paper';
export default class MaterialBottomTabView extends React.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "_getColor", ({
      route
    }) => {
      const {
        descriptors
      } = this.props;
      const descriptor = descriptors[route.key];
      const options = descriptor.options;

      if (this.context === 'dark' && options.tabBarColorDark) {
        return options.tabBarColorDark;
      } else if (options.tabBarColorLight) {
        return options.tabBarColorLight;
      } else {
        return options.tabBarColor;
      }
    });

    _defineProperty(this, "_getactiveColor", () => {
      let {
        activeColor,
        activeColorLight,
        activeColorDark
      } = this.props;

      if (this.context === 'dark' && activeColorDark) {
        return activeColorDark;
      } else if (activeColorLight) {
        return activeColorLight;
      } else {
        return activeColor;
      }
    });

    _defineProperty(this, "_getInactiveColor", () => {
      let {
        inactiveColor,
        inactiveColorLight,
        inactiveColorDark
      } = this.props;

      if (this.context === 'dark' && inactiveColorDark) {
        return inactiveColorDark;
      } else if (inactiveColorLight) {
        return inactiveColorLight;
      } else {
        return inactiveColor;
      }
    });

    _defineProperty(this, "_getBarStyle", () => {
      let {
        barStyle,
        barStyleLight,
        barStyleDark
      } = this.props;
      return [barStyle, this.context === 'dark' ? barStyleDark : barStyleLight];
    });

    _defineProperty(this, "_renderIcon", ({
      route,
      focused,
      color
    }) => {
      return this.props.renderIcon({
        route,
        focused,
        tintColor: color
      });
    });
  }

  _isVisible() {
    const {
      navigation,
      descriptors
    } = this.props;
    const {
      state
    } = navigation;
    const route = state.routes[state.index];
    const options = descriptors[route.key].options;
    return options.tabBarVisible;
  }

  render() {
    const {
      navigation,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      descriptors,
      ...rest
    } = this.props;

    const activeColor = this._getactiveColor();

    const inactiveColor = this._getInactiveColor();

    const barStyle = this._getBarStyle();

    const isVisible = this._isVisible();

    const extraStyle = isVisible === false ? {
      display: 'none',
      // When keyboard is shown, `position` is set to `absolute` in the library
      // This somehow breaks `display: 'none'`, so we explcitely override `position`
      position: undefined
    } : null;
    return /*#__PURE__*/React.createElement(BottomNavigation // Pass these for backward compaibility
    , _extends({}, rest, {
      activeColor: activeColor,
      inactiveColor: inactiveColor,
      renderIcon: this._renderIcon,
      barStyle: [barStyle, extraStyle],
      navigationState: navigation.state,
      getColor: this._getColor
    }));
  }

}

_defineProperty(MaterialBottomTabView, "contextType", ThemeContext);
