// react-navigation-drawer/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createDrawerNavigator", {
  enumerable: true,
  get: function () {
    return _createDrawerNavigator.default;
  }
});
Object.defineProperty(exports, "DrawerRouter", {
  enumerable: true,
  get: function () {
    return _DrawerRouter.default;
  }
});
Object.defineProperty(exports, "DrawerNavigatorItems", {
  enumerable: true,
  get: function () {
    return _DrawerNavigatorItems.default;
  }
});
Object.defineProperty(exports, "DrawerItems", {
  enumerable: true,
  get: function () {
    return _DrawerNavigatorItems.default;
  }
});
Object.defineProperty(exports, "DrawerSidebar", {
  enumerable: true,
  get: function () {
    return _DrawerSidebar.default;
  }
});
Object.defineProperty(exports, "DrawerView", {
  enumerable: true,
  get: function () {
    return _DrawerView.default;
  }
});
Object.defineProperty(exports, "DrawerGestureContext", {
  enumerable: true,
  get: function () {
    return _DrawerGestureContext.default;
  }
});
Object.defineProperty(exports, "DrawerProgressContext", {
  enumerable: true,
  get: function () {
    return _DrawerProgressContext.default;
  }
});
Object.defineProperty(exports, "NavigationDrawerState", {
  enumerable: true,
  get: function () {
    return _types.NavigationDrawerState;
  }
});
Object.defineProperty(exports, "NavigationDrawerProp", {
  enumerable: true,
  get: function () {
    return _types.NavigationDrawerProp;
  }
});
Object.defineProperty(exports, "NavigationDrawerOptions", {
  enumerable: true,
  get: function () {
    return _types.NavigationDrawerOptions;
  }
});
Object.defineProperty(exports, "NavigationDrawerConfig", {
  enumerable: true,
  get: function () {
    return _types.NavigationDrawerConfig;
  }
});
Object.defineProperty(exports, "NavigationDrawerRouterConfig", {
  enumerable: true,
  get: function () {
    return _types.NavigationDrawerRouterConfig;
  }
});
Object.defineProperty(exports, "NavigationDrawerScreenProps", {
  enumerable: true,
  get: function () {
    return _types.NavigationDrawerScreenProps;
  }
});
Object.defineProperty(exports, "NavigationDrawerScreenComponent", {
  enumerable: true,
  get: function () {
    return _types.NavigationDrawerScreenComponent;
  }
});
Object.defineProperty(exports, "DrawerContentComponentProps", {
  enumerable: true,
  get: function () {
    return _types.DrawerContentComponentProps;
  }
});
Object.defineProperty(exports, "DrawerLockMode", {
  enumerable: true,
  get: function () {
    return _types.DrawerLockMode;
  }
});
Object.defineProperty(exports, "DrawerIconProps", {
  enumerable: true,
  get: function () {
    return _types.DrawerIconProps;
  }
});
Object.defineProperty(exports, "DrawerLabelProps", {
  enumerable: true,
  get: function () {
    return _types.DrawerLabelProps;
  }
});
exports.DrawerActions = void 0;

var DrawerActions = _interopRequireWildcard(require("./routers/DrawerActions"));

exports.DrawerActions = DrawerActions;

var _createDrawerNavigator = _interopRequireDefault(require("./navigators/createDrawerNavigator"));

var _DrawerRouter = _interopRequireDefault(require("./routers/DrawerRouter"));

var _DrawerNavigatorItems = _interopRequireDefault(require("./views/DrawerNavigatorItems"));

var _DrawerSidebar = _interopRequireDefault(require("./views/DrawerSidebar"));

var _DrawerView = _interopRequireDefault(require("./views/DrawerView"));

var _DrawerGestureContext = _interopRequireDefault(require("./utils/DrawerGestureContext"));

var _DrawerProgressContext = _interopRequireDefault(require("./utils/DrawerProgressContext"));

var _types = require("./types");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
